### PR TITLE
Support bazel remote cache for builds

### LINKED ISSUE #89: Cache bazel artifacts in CI
I tried naively caching the whole bazel cache folder via the standard [Travis mechanism](https://docs.travis-ci.com/user/caching/#arbitrary-directories), but it always fails to restore it, even if I increase the timeout. I think the cache is too large and there are too many small files for Travis to unpack it in a reasonable time.
Another option could be to use a [Bazel remote cache](https://docs.bazel.build/versions/master/remote-caching.html#google-cloud-storage), which has the advantage that we could also share the same cache when building locally.

### PR DESCRIPTION
**Pull request by anghelcovici**
This will start using the cache from
https://storage.googleapis.com/oak-bazel-cache to store build artefacts.
In order to access it, a json key needs to exist either in the root
project folder: ./.oak_remote_cache_key.json or in the ENV var $BAZEL_GOOGLE_CREDENTIALS

This fixes #89

### CODE REVIEW

**Comment by anghelcovici** [comment]
This doesn't work for PRs as the BAZEL_GOOGLE_CREDENTIALS is not accessible from forks.

File: .gitignore [hunk 1]
```diff
@@ -15,6 +15,9 @@
 # emacs stuff
 *~
 
+# Ignore the bazel remote cache key file
```
Comment (tiziano88) [inline .gitignore]: Add a link to https://docs.bazel.build/versions/master/remote-caching.html#google-cloud-storage ?

Comment (tiziano88) [inline .gitignore]: End comments with `.` (and below).

File: .gitignore [hunk 2]
```diff
@@ -15,6 +15,9 @@
...
+.oak_remote_cache_key.json
```
Comment (tiziano88) [inline .gitignore]: Prefix with `/`.

File: scripts/build_server_docker [hunk 3]
```diff
@@ -4,5 +4,27 @@ set -o errexit
 set -o xtrace
 
 readonly SCRIPTS_DIR="$(dirname "$0")"
+readonly OAK_REMOTE_CACHE_KEY='./.oak_remote_cache_key.json'
+
+echo "TEST"
```
Comment (tiziano88) [inline scripts/build_server_docker]: Remove?

File: scripts/build_server_docker [hunk 4]
```diff
@@ -4,5 +4,27 @@ set -o errexit
...
+# Do we have a json key for the remote cache
+if [ ! -f $OAK_REMOTE_CACHE_KEY ]; then
```
Comment (tiziano88) [inline scripts/build_server_docker]: Prefer `[[` (see https://google.github.io/styleguide/shell.xml#Test,_%5B_and_%5B%5B ). Also quote variable. I think shellcheck should check some of these things?

File: scripts/build_server_docker [hunk 5]
```diff
@@ -4,5 +4,27 @@ set -o errexit
 set -o xtrace
 
 readonly SCRIPTS_DIR="$(dirname "$0")"
+readonly OAK_REMOTE_CACHE_KEY='./.oak_remote_cache_key.json'
+
+echo "TEST"
+# Do we have a json key for the remote cache
```
Comment (tiziano88) [inline scripts/build_server_docker]: (nit) JSON

Comment (tiziano88) [reply scripts/build_server_docker]: Also worth adding a link to https://docs.bazel.build/versions/master/remote-caching.html#google-cloud-storage around here.

File: .gitignore (hunk 1 above)
Comment (anghelcovici) [reply .gitignore]: None of the comments in .gitignore have a . :(

Comment (anghelcovici) [reply .gitignore]: Added.

Comment (anghelcovici) [reply .gitignore]: Done.

File: .gitignore (hunk 2 above)
Comment (anghelcovici) [reply .gitignore]: Done

File: scripts/build_server_docker (hunk 3 above)
Comment (anghelcovici) [reply scripts/build_server_docker]: Done

File: scripts/build_server_docker (hunk 5 above)
Comment (anghelcovici) [reply scripts/build_server_docker]: Done, both

File: scripts/build_server_docker (hunk 4 above)
Comment (anghelcovici) [reply scripts/build_server_docker]: Done. I don't think shellcheck picked it up. It might miss a rule?

**Comment by tiziano88** [comment]
I'm still not convinced this is working as intended, it seems we are still rebuilding protobuf and grpc over and over again: https://travis-ci.org/project-oak/oak/jobs/574389849 .

**Comment by tiziano88** [comment]
Ah my bad, this is explained in https://github.com/project-oak/oak/issues/89#issuecomment-512909557

**Pull request merged**